File: app/src/main/java/ru/org/adons/cblock/ui/viewmodel/AddViewModel.java
package ru.org.adons.cblock.ui.viewmodel;

import java.util.HashSet;
import java.util.List;

import javax.inject.Inject;

import ru.org.adons.cblock.data.BlockListModel;
import ru.org.adons.cblock.data.CallLogModel;
import ru.org.adons.cblock.model.CallLogItem;
import ru.org.adons.cblock.ui.view.AddFragment;
import ru.org.adons.cblock.utils.Logging;
import rx.Observable;

/**
 * Add Model View, provide data for {@link AddFragment}
 */

public class AddViewModel {

    @Inject CallLogModel callLogModel;
    @Inject BlockListModel blockListModel;

    public Observable<List<CallLogItem>> getCallLogList() {
        return callLogModel.getCallLogList()
                .doOnSubscribe(Logging.subscribe(this.getClass(), "getCallLogList"))
                .doOnUnsubscribe(Logging.unsubscribe(this.getClass(), "getCallLogList"));
    }

    public Observable<List<CallLogItem>> setBlockedItems(List<CallLogItem> items) {
        final HashSet<String> filter = new HashSet<>();
        return blockListModel.getBlockedPhones()
                .doOnSubscribe(Logging.subscribe(this.getClass(), "setBlockedItems"))
                .doOnUnsubscribe(Logging.unsubscribe(this.getClass(), "setBlockedItems"))
                .map(filter::addAll)
                .flatMap(phones -> Observable.from(items))
                .map(item -> {
                    if (filter.contains(item.phoneNumber())) {
                        item.isBlocked = true;
                    }
                    return item;
                })
                .toList();
    }

}

File: app/src/main/java/mobile/addons/cblock/ui/view/add/AddViewModel.kt
package mobile.addons.cblock.ui.view.add

import mobile.addons.cblock.app.BlockManager
import mobile.addons.cblock.data.CallLogItem
import mobile.addons.cblock.ext.logSubscribe
import mobile.addons.cblock.ext.logUnsubscribe
import mobile.addons.cblock.model.BlockListModel
import mobile.addons.cblock.model.CallLogModel
import rx.Observable
import javax.inject.Inject

/**
 * Add Model View, provide data for [AddFragment]
 */

class AddViewModel @Inject internal constructor() {

    @Inject lateinit var callLogModel: CallLogModel
    @Inject lateinit var blockListModel: BlockListModel
    @Inject lateinit var blockManager: BlockManager

    /**
     * Get list of incoming or missed calls and set isBlocked flag
     */
    fun getCallLogList(): Observable<List<CallLogItem>> = callLogModel
            .getCallLogList()
            .doOnSubscribe { logSubscribe("getCallLogList") }
            .doOnUnsubscribe { logUnsubscribe("getCallLogList") }
            .doOnNext { setBlockedNumbers(it) }

    fun addPhones(items: List<CallLogItem>) = blockManager
            .addPhones(blockListModel, items)

    // set isBlocked flag if phone already blocked, checkbox will disabled in UI
    private fun setBlockedNumbers(items: List<CallLogItem>): List<CallLogItem> {
        val blockList = with(blockListModel) { getBlockedPhones(getNumbersList()) }
        items.forEach { if (blockList.contains(it.phoneNumber)) it.isBlocked = true }
        return items
    }

}
